### PR TITLE
PLT-2899 adding clustering of app servers

### PR DESCRIPTION
**Pull request by coreyhulen**
#### Summary
Adds clustering of app servers

[related enterprise PR](https://github.com/mattermost/enterprise/pull/48)

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-2899

#### Checklist
- [X] Added or updated unit tests (required for all new features)
- [X] All new/modified APIs include changes to the drivers
- [X] Has enterprise changes (please link)
- [X] Has UI changes
- [X] Includes text changes and localization files updated
- [X] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: i18n/en.json [hunk 1]
```diff
@@ -2132,6 +2132,10 @@
     "translation": "Unable to open image. Image is too large."
   },
   {
+    "id": "ent.cluster.licence_disable.app_error",
+    "translation": "Clustering functionality disabled by current license. Please contact your system administrator about upgrading your enterprise license."
```
Comment (it33) [inline i18n/en.json]: Propose changing: 

```
Clustering functionality disabled by current license. Please contact your system administrator about upgrading your enterprise license.
```

to

```
Clustering functionality is not available under your current Enterprise Edition license. Contact your system administrator to upgrade your license.
```

File: i18n/en.json [hunk 2]
```diff
@@ -2132,6 +2132,10 @@
...
+  },
+  {
     "id": "ent.compliance.licence_disable.app_error",
     "translation": "Compliance functionality disabled by current license. Please contact your system administrator about upgrading your enterprise license."
```
Comment (it33) [inline i18n/en.json]: Propose changing

```
Compliance functionality disabled by current license. Please contact your system administrator about upgrading your enterprise license.
```

to 

```
Compliance functionality is not available under your current Enterprise Edition license. Please contact your system administrator about upgrading your license.
```

File: i18n/en.json [hunk 3]
```diff
@@ -4638,5 +4642,41 @@
...
     "id": "web.watcher_fail.error",
     "translation": "Failed to add directory to watcher %v"
+  },
+  {
+    "id": "ent.cluster.starting.info",
+    "translation": "Cluster internode communication is listening on %v with hostname=%v id=%v"
+  },
+  {
+    "id": "ent.cluster.config_changed.info",
+    "translation": "Cluster configuration has changed for id=%v.  Attempting to restart cluster service.  To ensure the cluster is configured correctly you should not rely on this restart because we detected a core configuration change."
```
Comment (it33) [inline i18n/en.json]: @coreyhulen maybe we can discuss this message? Not totally sure what it means...

File: webapp/components/admin_console/cluster_settings.jsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        if (AdminStore.getClusterId()) {
+            configLoadedFromCluster = (
+                <div
+                    style={{marginBottom: '10px'}}
+                    className='alert alert-warning'
+                >
+                    <i className='fa fa-warning'></i>
+                    <FormattedMessage
+                        id='admin.cluster.loadedFrom'
+                        defaultMessage='This config file was loaded from Node ID "{clusterId}".  If you are accessing the System Console through a load balancer you may experience issues.'
```
Comment (it33) [inline webapp/components/admin_console/cluster_settings.jsx]: @coreyhulen wondering if we could discuss this message to say a little more about the situation?

File: webapp/components/admin_console/cluster_settings.jsx [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        if (this.state.showWarning) {
+            warning = (
+                <div
+                    style={{marginBottom: '10px'}}
+                    className='alert alert-warning'
+                >
+                    <i className='fa fa-warning'></i>
+                    <FormattedMessage
+                        id='admin.cluster.should_not_change'
+                        defaultMessage='WARNING: these settings may not sync with the other servers in the cluster.  You must verify the config files are identical and restart the cluster.  We will attempt to save the settings across the servers.  Changing from disabled to enabled will not start the High Availability inter-node communication and you will need to modify the config.json on the other servers and restart Mattermost.  Please see the documentation on how to enable and join or remove a server to the cluster. If you are accessing the System Console through a load balancer you may experience issues.  Yoy may need to access the admin console for each server in the cluster by accessing the internal/private ip address of the server.'
```
Comment (it33) [inline webapp/components/admin_console/cluster_settings.jsx]: @coreyhulen let's mind meld a bit on this message when you have time?

File: webapp/i18n/en.json [hunk 6]
```diff
@@ -514,6 +514,24 @@
   "admin.saml.usernameAttrTitle": "Username Attribute:",
   "admin.saml.verifyDescription": "When true, Mattermost verifies that the signature sent from the SAML Response matches the Service Provider Login URL",
   "admin.saml.verifyTitle": "Verify Signature:",
+  "admin.cluster.loadedFrom": "This config file was loaded from Node ID {clusterId}.  If you are accessing the System Console through a load balancer you may experience issues.",
```
Comment (it33) [inline webapp/i18n/en.json]: @coreyhulen let's mind meld a bit on this message when you have time?

**Comment by enahum** [comment]
needs rebase

File: webapp/i18n/en.json [hunk 7]
```diff
@@ -540,6 +540,24 @@
...
+  "admin.cluster.noteDescription": "Changing properties in this section will require a server restart before taking effect and will set the System Console to read-only mode.  You must enable this via the configuration file.",
+  "admin.cluster.enableTitle": "Enable Login With High Availability:",
+  "admin.cluster.enableDescription": "When true, Mattermost will run in High Availability mode. Please see <a href=\"http://docs.mattermost.com/deployment/cluster.html\" target=\"_blank\">documentation</a> to learn more about configuring High Availability for Mattermost.",
+  "admin.cluster.interNodeListenAddressTitle": "Inter-Node Listen Address:",
+  "admin.cluster.interNodeListenAddressEx": "Ex \":8075\"",
+  "admin.cluster.interNodeListenAddressDesc": "The address the server will listen on for communicating with other servers.",
+  "admin.cluster.interNodeUrlsTitle": "Inter-Node URLs:",
+  "admin.cluster.interNodeUrlsEx": "Ex \"http://10.10.10.30, http://10.10.10.31\"",
+  "admin.cluster.interNodeUrlsDesc": "The internal/private URLs of all the Mattermost server seperated by commas.",
```
Comment (esethna) [inline webapp/i18n/en.json]: Fixed Typos: "The internal/private URLs of all the Mattermost servers separated by commas."

File: webapp/i18n/en.json [hunk 8]
```diff
@@ -540,6 +540,24 @@
...
+  "admin.cluster.should_not_change": "WARNING: these settings may not sync with the other servers in the cluster.  You must verify the config files are identical and restart the cluster.  We will attempt to save the settings across the servers.  Changing from disabled to enabled will not start the High Availability inter-node communication and you will need to modify the config.json on the other servers and restart Mattermost.  Please see the documentation on how to enable and join or remove a server to the cluster. If you are accessing the System Console through a load balancer you may experience issues.  Yoy may need to access the admin console for each server in the cluster by accessing the internal/private ip address of the server.",
+  "admin.cluster.noteDescription": "Changing properties in this section will require a server restart before taking effect and will set the System Console to read-only mode.  You must enable this via the configuration file.",
+  "admin.cluster.enableTitle": "Enable Login With High Availability:",
```
Comment (esethna) [inline webapp/i18n/en.json]: Propose: `Enable High Availability Mode`

File: webapp/i18n/en.json [hunk 9]
```diff
@@ -540,6 +540,24 @@
...
+  "admin.cluster.loadedFrom": "This config file was loaded from Node ID {clusterId}.  If you are accessing the System Console through a load balancer you may experience issues.",
+  "admin.cluster.should_not_change": "WARNING: these settings may not sync with the other servers in the cluster.  You must verify the config files are identical and restart the cluster.  We will attempt to save the settings across the servers.  Changing from disabled to enabled will not start the High Availability inter-node communication and you will need to modify the config.json on the other servers and restart Mattermost.  Please see the documentation on how to enable and join or remove a server to the cluster. If you are accessing the System Console through a load balancer you may experience issues.  Yoy may need to access the admin console for each server in the cluster by accessing the internal/private ip address of the server.",
+  "admin.cluster.noteDescription": "Changing properties in this section will require a server restart before taking effect and will set the System Console to read-only mode.  You must enable this via the configuration file.",
```
Comment (esethna) [inline webapp/i18n/en.json]: Can we be more clear about when the System Console is set to "read-only mode" vs read and write? Is the following accurate?:

"Changing properties in this section will require a server restart before taking effect. When High Availability mode is disabled, the System Console is set to read-only and can only be enabled in the configuration file.

File: webapp/i18n/en.json [hunk 10]
```diff
@@ -540,6 +540,24 @@
   "admin.saml.usernameAttrTitle": "Username Attribute:",
   "admin.saml.verifyDescription": "When true, Mattermost verifies that the signature sent from the SAML Response matches the Service Provider Login URL",
   "admin.saml.verifyTitle": "Verify Signature:",
+  "admin.cluster.loadedFrom": "This config file was loaded from Node ID {clusterId}.  If you are accessing the System Console through a load balancer you may experience issues.",
```
Comment (esethna) [inline webapp/i18n/en.json]: "config" > "configuration". Should we link to the documentation where we would have a troubleshooting guide?

`This configuration file was loaded from Node ID {clusterId}. Please see the Troubleshooting Guide in our <a href=\"http://docs.mattermost.com/deployment/cluster.html\" target=\"_blank\">documentation</a> if you are experiencing issues accessing the System Console through a load balancer.`

File: webapp/i18n/en.json [hunk 11]
```diff
@@ -540,6 +540,24 @@
...
+  "admin.cluster.should_not_change": "WARNING: these settings may not sync with the other servers in the cluster.  You must verify the config files are identical and restart the cluster.  We will attempt to save the settings across the servers.  Changing from disabled to enabled will not start the High Availability inter-node communication and you will need to modify the config.json on the other servers and restart Mattermost.  Please see the documentation on how to enable and join or remove a server to the cluster. If you are accessing the System Console through a load balancer you may experience issues.  Yoy may need to access the admin console for each server in the cluster by accessing the internal/private ip address of the server.",
```
Comment (esethna) [inline webapp/i18n/en.json]: Some of these sentences seem redundant. Please review for accuracy but I propose:

`WARNING: These settings may not sync with the other servers in the cluster. High Availability inter-node communication will not start until you modify the config.json to be identical on all servers and restart Mattermost. The System Console for each server in the cluster is accessible by the internal/private IP address.  Please see the <a href=\"http://docs.mattermost.com/deployment/cluster.html\" target=\"_blank\">documentation</a> on how to add or remove a server from the cluster. If you are accessing the System Console through a load balancer and experiencing issues, please see the Troubleshooting Guide in our <a href=\"http://docs.mattermost.com/deployment/cluster.html\" target=\"_blank\">documentation</a>. `

Also, do the IT admins need to access the system console for each server individually or can they make all the necessary modifications via the config files? If they can do it from the config files I'd propose removing the sentence `The System Console for each server in the cluster is accessible by the internal/private IP address.`

**Comment by crspeller** [comment]
+1 Other than the nitpick in the EE repo.

File: webapp/i18n/en.json (hunk 7 above)
Comment (coreyhulen) [inline webapp/i18n/en.json]: fixed

File: webapp/i18n/en.json (hunk 8 above)
Comment (coreyhulen) [inline webapp/i18n/en.json]: fixed

File: webapp/i18n/en.json (hunk 9 above)
Comment (coreyhulen) [inline webapp/i18n/en.json]: fixed, tweaked a little

File: webapp/i18n/en.json (hunk 10 above)
Comment (coreyhulen) [inline webapp/i18n/en.json]: fixed

File: webapp/i18n/en.json (hunk 11 above)
Comment (coreyhulen) [inline webapp/i18n/en.json]: fixed, I removed the line as suggested

**Comment by coreyhulen** [comment]
@esethna fixed strings as suggested (with a few tweaks)

**Comment by esethna** [comment]
PM +1 on the strings

**Comment by enahum** [comment]
+1

**Comment by crspeller** [comment]
@enahum What PR is this awaiting?

**Comment by enahum** [comment]
The EE PR is still marked as `DO NOT MERGE`

**Pull request merged**